packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,400 @@
+  ///
+  /// The range that is being replaced can either grow or shrink based on the
+  /// given replacement text.
+  ///
+  /// A replacement can occur in cases such as auto-correct, suggestions, and
+  /// when a selection is replaced by a single character.
+  /// {@endtemplate}
+  replacement,
+
+  /// {@template flutter.services.TextEditingDeltaNonTextUpdate}
+  /// The delta is not modifying the text. There are potentially selection and
+  /// composing region updates in the delta that still need to be applied to your
+  /// text model.
+  ///
+  /// A situation where this delta would be created is when dragging the selection
+  /// handles. There are no changes to the text, but there are updates to the selection
+  /// and potentially the composing region as well.
+  /// {@endtemplate}
+  nonTextUpdate,
+}
+
+/// A structure representing a granular change that has occurred to the editing
+/// state as a result of text editing.
+abstract class TextEditingDelta {
+  /// Creates a delta for a given change to the editing state.
+  ///
+  /// {@template flutter.services.TextEditingDelta}
+  /// The [oldText], [deltaText], [deltaRange], [selection], and [composing]
+  /// arguments must not be null.
+  /// {@endtemplate}
+  const TextEditingDelta({
+    required this.oldText,
+    this.deltaText = '',
+    this.deltaRange = TextRange.empty,
+    required this.selection,
+    required this.composing,
+  }) : assert(oldText != null),
+       assert(deltaText != null),
+       assert(deltaRange != null),
+       assert(selection != null),
+       assert(composing != null);
+
+  /// Creates an instance of this class from a JSON object by inferring the
+  /// type of delta based on values sent from the engine.
+  factory TextEditingDelta.fromJSON(Map<String, dynamic> encoded) {
+    final String oldText = encoded['oldText'] as String;
+    final int start = encoded['deltaStart'] as int;
+    final int end = encoded['deltaEnd'] as int;
+    final String tb = encoded['deltaText'] as String;
+    const int tbStart = 0;
+    final int tbEnd = tb.length;
+
+    // This delta is explicitly a non text update.
+    final bool isNonTextUpdate = start == -1 && start == end;
+    final TextRange newComposing = TextRange(
+      start: encoded['composingBase'] as int? ?? -1,
+      end: encoded['composingExtent'] as int? ?? -1,
+    );
+    final TextSelection newSelection = TextSelection(
+      baseOffset: encoded['selectionBase'] as int? ?? -1,
+      extentOffset: encoded['selectionExtent'] as int? ?? -1,
+      affinity: _toTextAffinity(encoded['selectionAffinity'] as String?) ??
+          TextAffinity.downstream,
+      isDirectional: encoded['selectionIsDirectional'] as bool? ?? false,
+    );
+
+    if (isNonTextUpdate) {
+      return TextEditingDeltaNonTextUpdate(
+        oldText: oldText,
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+
+    final String textStart = oldText.substring(0, start);
```

Comment:
The chunk of logic starting here to the end of the method is prone to missing edge cases, especially if it needs to be maintained. You should add documentation here explaining the specific conditions each type of delta must satisfy to try to make this clear what it should be implementing.

From reading this code, it is hard to know if any edge cases were missed or not.

Reply:
Added an explanation of the conditions.